### PR TITLE
Doc - Release using channels

### PR DESCRIPTION
**Pull request by popod**
todo:
- [ ] improve documentation
- [ ] review and fix english typos

### CODE REVIEW

File: docs/tutorials/release-using-channels.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+> !!! This documentation is in "beta" and needed to be tested !!!
+
+# Release using channels / Auto-updates with channels
```
Comment (develar) [inline docs/tutorials/release-using-channels.md]: Please in the future use form  "Header Must Use Capital Case".

File: docs/tutorials/release-using-channels.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+> Note: `allowDowngrade` is automatically set to `true` when `generateUpdatesFilesForAllChannels = true`, so you doesn't need to set it.
+
+All you have to do to release using channels is to define the channel in the version tag of the `package.json`. Add "-beta" or "-alpha" (nothing for stable) to automatically build for the related channel.
+
+
+### Your application
+All you need to do here is to define which channel the user will receive.
+
+`autoUpdater.setFeedURL({url: 'https://www.yoursite.com', channel: 'beta'});`
```
Comment (develar) [inline docs/tutorials/release-using-channels.md]: Now user can set `channel`, `setFeedURL` not required, right?

**Comment by kwent** [comment]
One scary [section](https://github.com/electron-userland/electron-builder/blob/e409c0e9a4542204ef3d16048700470edb4b02df/docs/auto-update.md#quick-setup-guide) to me was:

```
Do not call setFeedURL. electron-builder automatically creates app-update.yml file for you on build in the resources (this file is internal, you don't need to be aware of it).
```

Might want to modify this as well :)

File: docs/tutorials/release-using-channels.md (hunk 2 above)
Comment (kwent) [reply docs/tutorials/release-using-channels.md]: I'm confused. Where do we set `channel` ?

Comment (kwent) [reply docs/tutorials/release-using-channels.md]: Oh this: https://github.com/electron-userland/electron-builder/blob/ba4809aeb3bd3b27f66ee684c3e6e8ab0cf55186/packages/electron-updater/src/AppUpdater.ts#L73. Needs to be documented

File: docs/tutorials/release-using-channels.md (hunk 1 above)
Comment (popod) [reply docs/tutorials/release-using-channels.md]: Thanks for the comment, I've change this.

File: docs/tutorials/release-using-channels.md (hunk 2 above)
Comment (popod) [reply docs/tutorials/release-using-channels.md]: @kwent and @develar Thanks for feedback. I've remove the "old" `setFeedURL` and replace it with `autoUpdater.channel()`. I will add the doc for `channel()` too.

File: docs/tutorials/release-using-channels.md [hunk 3]
```diff
@@ -36,9 +37,9 @@ All you have to do to release using channels is to define the channel in the ver
 
 
 ### Your Application
-All you need to do here is to define which channel the user will receive.
+All you need to do here is to define which channel the user will receive with:
 
-`autoUpdater.channel('beta')`
+`autoUpdater.channel('beta')` (see [the documentation here](../auto-update.md#module_electron-updater.AppUpdater+channel))
```
Comment (kwent) [inline docs/tutorials/release-using-channels.md]: I have been doing `autoUpdater.channel = 'beta'`. Is this valid as well ?

Comment (popod) [reply docs/tutorials/release-using-channels.md]: @kwent my bad.. channel('beta') is false. I will correct this.

**Comment by popod** [comment]
@develar and @kwent what do you think about this current "small tutorial" ? Have I miss something ?

File: docs/tutorials/release-using-channels.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
+<a name="release_using_channels"></a>
+# Release Using Channels / Auto-Updates With Channels
+
+## Description
+Channels are useful to distribute "beta" or "alpha" releases of your application to a chosen set of users. This allows to test an application before release it as "stable".
+
+Users which receive "beta" version will get "stable" versions too. Otherwise, users who don't want "beta" will only get "stable" releases.
+
+They are tree channels level ordered by stability:
```
Comment (rahilsondhi) [inline docs/tutorials/release-using-channels.md]: typo. i think you meant to say "There are three channels, ordered by stability"

Comment (popod) [reply docs/tutorials/release-using-channels.md]: Thanks ! Yes this is exactly what I'm trying to say ;)

**Comment by kwent** [comment]
LGTM.

One thing i saw is if i set channel to `stable` the library was looking for `stable-mac.yml`. (I imagine same thing for windows and linux) but the `packager` is creating `latest-mac.yml`. Setting the channel to `latest` instead `stable` fixed it in my case. Not sure if the `autoupdater` or the `packager` is wrong but something is not in phase.

Might not be the best place and might be worst opening an issue instead tho. cc @develar

**Comment by popod** [comment]
@kwent thanks: this is an error! The channels to use are `latest`, `beta` and `alpha`, not `stable`. I think the easy way is that I use `latest` in place of `stable` in this doc. But `stable` seems to be more intuitive..

@develar do you see an other possibility ? Like as example set the channel to `latest` both when the user set it to `stable` and `latest` ?

>(curiously you can set the channel to any string value as well...). Not sure if the autoupdater or the packager is wrong but something is not in phase.

And do you think that this is useful to limit the channel to `latest`, `beta` and `alpha` ?

**Comment by kwent** [comment]
Maybe we can alias stable to latest ?

**Comment by develar** [comment]
> And do you think that this is useful to limit the channel to latest, beta and alpha ?

No, user should be able to use whatever he want. But! Where do you set? updater.channel or electron-builder publish config channel?

**Comment by develar** [comment]
Obviously, that you cannot set updater.channel to `stable` if you use `generateUpdatesFilesForAllChannels`. 

`stable` in our case it is `latest`. But tutorial has a lot of usages of word `stable` :) And actually, to set  "stable" channel (as tutorial suggests), you need to set it to `null` or `latest`.

**Comment by popod** [comment]
@develar as I wrote in my latest comment, do you think that it is possible to do this:

When doing `autoUpdater.channel = 'stable'`, in the setter function, change `stable` by `latest` ? As is, the tutorial will be correct. I think that the word "stable" is more logic than "latest" for a "user point of view" and according to the tutorial description.

**Comment by develar** [comment]
Yeach... you are right. `latest` is not clear. Because actually, `beta` will point to stable release if for now stable release is the latest.

> all application releases use the "stable" channel.

Well... no channel means that channels are not used and default name of update file is `latest` :)

Need to think about it. We cannot map `stable` to `latest` in the electron-updater, because electron-updater is not aware of `generateUpdatesFilesForAllChannels`. And electron-builder (if generateUpdatesFilesForAllChannels is set to default false) will generate `stable` update file instead of `latest`. I do not want to add this magic transformation.

Possible solution — if `generateUpdatesFilesForAllChannels` is set, `stable` will be generated instead of `latest` (and logic `getDefaultChannelName` of electron-updater will be not used because we will set channel by default explicitly to `stable`).

**Comment by popod** [comment]
Thanks for your reply @develar ! I understand and agree with the fact that we should no do "magic transformation" :)

I think that the auto-updater should be "compatible" if we use `generateUpdatesFilesForAllChannels` or not. So we need to keep the `latest` label. I think the easy way to solve this problem is that I update the tutorial with `latest` in place of `stable`.

I will do it tomorrow. If you have an other solution, let me know ;)

**Comment by jake** [comment]
Been loosely following this thread, and have actually got this multi-channel support working in my app. So, thanks for these WIP docs!

Isn't the feature already complete? Are we discussing improving how it works, or just documenting the way it currently works?

At the risk of getting off topic... I was definitely annoyed and disappointed that I had to use these magic variable names ("alpha" -> "beta" -> "latest"). For my workflow, I want "canary" -> "preview" -> "release". 

I think hardcoding these release channel names is a very weird idea. Not only does it limit functionality, it's less clear how they relate. If you let the user define the flow of channels themselves, they can customize whatever (and however many) make sense for their application, and it also will likely make more sense overall since they set up the relationships themselves.

There should be a new concept introduced in this release channels feature, one of channel order. I think in the build settings, defining an ordered array would be great.

Maybe, if you wanted to keep a sane default around, you could make `channels: true` expand to:

```
channels:
  - alpha
  - beta
  - latest
```

Or, if you want something else as I'd like to you could define your own array:

```
channels:
  - canary
  - preview
  - release
```

If `generateUpdatesFilesForAllChannels` is set, the same logic that exists today is followed, where anything _before_ the currently-being-built-to-channel in that array gets pushed the release as well. (e.g., if I was building "preview" it would also push to "canary").

Also, FWIW, `generateUpdatesFilesForAllChannels` is also a very confusingly named parameter. It doesn't generate updates for _all_ channels. It controls whether or not to publish the update to any channel "before" the currently built to channel.

**Comment by popod** [comment]
@jake Thanks for your feedback !

>Isn't the feature already complete? Are we discussing improving how it works, or just documenting the way it currently works?

Here, I think we should only "documenting the way it currently works". But we think about improving it too :)

Auto update with channel is available since a few month and no documentation existe about it until now.. As we discuss here: https://github.com/electron-userland/electron-builder/issues/1182#issuecomment-324944370, I already have suggest to let the user define it's own "channel name". But we kept things simple for the first working implementation.

I think that we should keep the "stable, beta, alpha" channels by default to provide an "easy to use" functionality, but allow advanced users to change this..

I don't know what think @develar about that, but I think we could open a new issue for let the users chose their "channels names" and the related order. I really think this is a pretty feature !

**Comment by develar** [comment]
> Isn't the feature already complete? 

Complete.

> generateUpdatesFilesForAllChannels is also a very confusingly named parameter. It doesn't generate updates for all channels

Well, this option does exactly what name means — generates _update info files_ for all channels :) As update info files is generated and uploaded, your users get updates.

@jake Do you realise and understand, that `canary` update info file will contain info about ** release**  channel update when do you release a new version of app? e.g. after canary 20.3.0 you release 20.40.30 and so, all users of `canary` or `preview ` channels must get updates from `release` channel.

> For my workflow, I want "canary" -> "preview" -> "release".

@jake Do you use semver or not? I mean — `canary` Please provide example of `canary` version number.

**Comment by develar** [comment]
> I think hardcoding these release channel names is a very weird idea. Not only does it limit functionality, it's less clear how they relate. 

BTW, right now nothing stops you :) But it is your responsibility to copy/create/upload such update info files. And thanks a lot to @popod, it is not good — electron-builder should help you to avoid manual uploading/copying.

**Comment by jake** [comment]
Sorry, I don't quite follow what you're asking. I also just wanted to reiterate I'm just trying to be helpful by providing some real world feedback :)

I don't use semver, as I think it's very non-human-friendly. I try to build software for humans, not robots. I understand its value in internal library world, but Electron is for end users. I'm not sure how relevant it is to this discussion, but here's our versioning strategy. Ideally the names would be what I described earlier. FWIW, Framer.js also uses a similar [whole-version-number](https://framer.com/updates/) strategy.

<details>
<summary>my non-standard versioning strategy</summary>
Per the release channel docs ([issue](https://github.com/electron-userland/electron-builder/issues/1182), [pull](https://github.com/electron-userland/electron-builder/pull/2595)), there are three release channels: Alpha, Beta, Stable.

Build artifacts have their version number (and thus their release channel) in the filename.

Once you download a release from a channel, you are subscribed to that channel. You can easily switch by downloading a release from a different channel.

Conceptually, we are working on `3.0.0` within the Alpha and Beta channels, and finally releasing it through the Stable channel. Once `3.0.0` is released on Stable we start work on `4.0.0` within the Alpha channel and repeat the workflow.

### Alpha
Rolling time-based snapshots. Version is next potential for "Stable" channel. Internal only, bleeding edge. Commonly broken. Possibly CI-driven every commit.

- `3.0.0-alpha.20180218.093433`
- `3.0.0-alpha.20180218.094454`
- `3.0.0-alpha.20180218.112235`

Note: Subscribers of Alpha also receives updates published to Beta and Stable.

### Beta
Latest snapshot promoted to Beta channel for consideration. Limited pool of external testers. *Supposed* to work and be ready for public consumption.

- `3.0.0-beta`

In the event a beta release has an issue, we issue hotfixes to the Beta channel.

- `3.0.0-beta.1`
- `3.0.0-beta.2`
- `3.0.0-beta.3`

Note: we'd also be issuing additional releases on the Snapshot channel, e.g. `3.0.0-snapshot.TIMESTAMP`, and once we solved the issue issuing new Beta channel releases.

Note: Subscribers of Beta also receives updates published to Stable.

Consider: TIMESTAMP on hotfix releases?

- `3.0.0-beta.20180218.112424`
- `3.0.0-beta.20180218.125310`
- `3.0.0-beta.20180219.152239`

### Stable

The publicly consumed versions. Guaranteed to work, as they passed through Snapshot and Beta channels. Only ever whole major versions with zeroed minor and patch versions.

- `1.0.0`
- `2.0.0`
- `3.0.0`

Note: Subscribers of Stable only receive updates to Stable.
</details>

---

In your example, is it not still pushing to both the currently selected channel and any channels it supersedes?

- If I'm building `latest` it pushes to `latest`, `beta`, and `alpha`
- If I'm building `beta`, it pushes to `beta` and `alpha`
- If I'm building `alpha`, it pushes to `alpha`

Isn't this always the case? Release channels cascade down. You never push a channel release "upstream"-- you wouldn't build an `alpha` release and push it to `latest`. However, you'd want people subscribed to a "downstream" channel to get anything upstream.

---

> Well, this option does exactly what name means — generates update info files for all channels :) As update info files is generated and uploaded, your users get updates.

How so? I have this option enabled, and when I publish an "alpha" build my "latest" release file is not regenerated, or even re-uploaded as far as I can tell. Doesn't this option control the feature that pushes releases to downstream channels in addition to the selected channel?

> right now nothing stops you

Are you proposing I fork `electron-builder` and change those hardcoded values?

**Comment by develar** [comment]
> Are you proposing I fork electron-builder and change those hardcoded values?

These values are not harcoded — you can use any channel name as you want (but in this case electron-builder cannot help you to upload/regenerate corresponding update info files).

> when I publish an "alpha" build my "latest" release file is not regenerated, or even re-uploaded as far as I can tell.

Yes. Because `latest` it is stable, and as you build `alpha`, only `alpha` is regenerated and uploaded.

Ok... so, as far I see from your example versions, you use pre-release version tags `alpha` and `beta`. But... you name (want to think in these terms, want to name it so) channels as `canary` and `preview, right?

**Comment by jake** [comment]
> These values are not harcoded — you can use any channel name as you want (but in this case electron-builder cannot help you to upload/regenerate corresponding update info files).

Oh, I see. I guess I'm somewhat conflating arbitrary release channels with the "sorted" release channels as I've been describing (e.g. canary -> preview -> stable).

The `alpha` -> `beta` -> `latest` release hierarchy _is_ hardcoded though, right?

> you use pre-release version tags alpha and beta. But... you name (want to think in these terms, want to name it so) channels as canary and preview, right?

Pretty much! I'm fighting against `npm`'s  annoyingly-hardcoded semver requirements. If I was just setting up a system without any limitations, I'd just use e.g. version `v3-canary` for the "canary" channel, `v3-preview` for "preview", and `v3` for "release".

Anyway, if I can be of any help testing the future fully-customizable release channel hierarchy I'm game!

**Comment by jake** [comment]
Oh, just to clarify:

Have "latest" **always** mean "ready for production" _unless you've defined your own release channel list_. In which case, the first entry in your list represents "ready for production", and it's whatever name you've chosen for it.

**Comment by develar** [comment]
> Then it's easy to do channel[0] to find the most-stable

Hmm... currently, we detect it using prerelease component of current app version. If no pre release component — only Most-stable channel file will be generated.

**Comment by kwent** [comment]
@popod @develar I think we need to fix the `stable` / `latest` ambiguous and this should be good to merge.

**Comment by popod** [comment]
@develar @kwent I've change the documentation by using "latest" in place of "stable". I think this is clear enough to deploy.

**Comment by kwent** [comment]
LGTM

**Comment by develar** [comment]
@popod Thanks a lot. The only reason why this awesome addition is not yet merged — I want to do some changes to address @jake suggestions.

**Comment by JsseL** [comment]
For those looking for a **[link to the docs](https://www.electron.build/tutorials/release-using-channels)**.

**Pull request merged**